vms/platformvm/utxo/camino_locked.go

```diff
@@ -0,0 +1,665 @@
+			amountToBurn := math.Min(
+				fee-amountBurned, // Amount we still need to burn
+				remainingValue,   // Amount available to burn
+			)
+			amountBurned += amountToBurn
+			remainingValue -= amountToBurn
+		}
+
+		// Stake any value that should be staked
+		amountToStake := math.Min(
+			amount-amountStaked, // Amount we still need to stake
+			remainingValue,      // Amount available to stake
+		)
+		amountStaked += amountToStake
+		remainingValue -= amountToStake
+
+		if lockIDs.IsLocked() {
+			// Add the input to the consumed inputs
+			ins = append(ins, &avax.TransferableInput{
+				UTXOID: utxo.UTXOID,
+				Asset:  avax.Asset{ID: h.ctx.AVAXAssetID},
+				In: &lock.LockedIn{
+					LockIDs:        lockIDs,
+					TransferableIn: in,
+				},
+			})
+		} else {
+			// Add the input to the consumed inputs
+			ins = append(ins, &avax.TransferableInput{
+				UTXOID: utxo.UTXOID,
+				Asset:  avax.Asset{ID: h.ctx.AVAXAssetID},
+				In:     in,
+			})
+
+		}
+
+		if amountToStake > 0 {
+			// Some of this input was put for staking
+			outs = append(outs, &avax.TransferableOutput{
+				Asset: avax.Asset{ID: h.ctx.AVAXAssetID},
+				Out: &lock.LockedOut{
+					LockIDs: lockIDs.Lock(appliedLockState),
+					TransferableOut: &secp256k1fx.TransferOutput{
+						Amt:          amountToStake,
+						OutputOwners: innerOut.OutputOwners,
+					},
+				},
+			})
+		}
+
+		// This input had extra value, so some of it must be returned
+		if remainingValue > 0 {
+			if lockIDs.IsLocked() {
+				outs = append(outs, &avax.TransferableOutput{
+					Asset: avax.Asset{ID: h.ctx.AVAXAssetID},
+					Out: &lock.LockedOut{
+						LockIDs: lockIDs,
+						TransferableOut: &secp256k1fx.TransferOutput{
+							Amt:          remainingValue,
+							OutputOwners: innerOut.OutputOwners,
+						},
+					},
+				})
+			} else {
+				outs = append(outs, &avax.TransferableOutput{
+					Asset: avax.Asset{ID: h.ctx.AVAXAssetID},
+					Out: &secp256k1fx.TransferOutput{
+						Amt: remainingValue,
+						OutputOwners: secp256k1fx.OutputOwners{
+							Locktime:  0,
+							Threshold: 1,
+							Addrs:     []ids.ShortID{changeAddr},
+						},
+					},
+				})
+			}
+		}
+
+		// Add the signers needed for this input to the set of signers
+		signers = append(signers, inSigners)
+	}
+
+	if amountBurned < fee || amountStaked < amount {
+		return nil, nil, nil, errNotEnoughBalance
+	}
+
+	avax.SortTransferableInputsWithSigners(ins, signers) // sort inputs and keys
+	avax.SortTransferableOutputs(outs, txs.Codec)        // sort outputs
+
+	return ins, outs, signers, nil
+}
+
+func (h *handler) Unlock(
+	state state.Chain,
+	lockTxIDs []ids.ID,
+	removedLockState lock.LockState,
+) (
+	[]*avax.TransferableInput, // inputs
+	[]*avax.TransferableOutput, // outputs
+	error,
+) {
+	if removedLockState != lock.LockStateBonded && removedLockState != lock.LockStateDeposited {
+		return nil, nil, errInvalidTargetLockState
+	}
+
+	lockTxAddresses := make(map[ids.ID]*ids.ShortSet, len(lockTxIDs))
```

Comment:
Better collect into a single map (make(map[ids.ID]*ids.ShortID) and skip the second loop

Reply:
even if state.LockedUTXOs will return only utxos matching lockTxID ?

Reply:
Right, my concern is that we could have lot of duplicate addesses, so we should think about passing a map[txID] and a merged map of addresses to GetLockedUTXOs.